### PR TITLE
[Catalog] covered product ViewModel AddToCompareAvailability by Unit Test

### PR DESCRIPTION
**Pull request by srsathish92**
### Description (*)
[Catalog] covered product ViewModel **AddToCompareAvailability** by Unit Test
Path: `app/code/Magento/Catalog/ViewModel/Product/Checker/AddToCompareAvailability.php`

### Manual testing scenarios (*)
`vendor/bin/phpunit -c dev/tests/unit/phpunit.xml.dist app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php`

### Questions or comments
If any please let me know your feedback 

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+    /**
+     * @var AddToCompareAvailability
+     */
+    private $viewModel;
+
+    /**
+     * @var StockConfigurationInterface|MockObject 
+     */
+    protected $stockConfiguration;
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php]: Add suffix `Mock` to use `stockConfigurationMock` as a name, it's clear that we're facing Mock, then.

File: app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+        $productMock->expects($this->any())
+            ->method('getQuantityAndStockStatus')
+            ->willReturn($isInStock);
+
+        $this->stockConfiguration->expects($this->any())
+            ->method('isShowOutOfStock')
+            ->willReturn($isShowOutOfStock);
+
+        return $this->viewModel->isAvailableForCompare($productMock);
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php]: Why are you returning anything from the test?

File: app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+     * @inheritdoc
+     */
+    protected function setUp(): void
+    {
+
+        $this->stockConfiguration =
+            $this->getMockBuilder(StockConfigurationInterface::class)
+            ->getMock();
+
+        $this->viewModel = $this->getObjectManager()->getObject(
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php]: If you are using ObjectManager only once - in setUp - you can use local variable for ObjectManager.

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php (hunk 2 above)
Comment (srsathish92) [reply app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php]: @lbajsarowicz Thanks for feedback. Actually `isShowOutOfStock` returns bool, So I passed data from data provider with `true` & `false`. Also `ScopeConfigInterface` not injected in original file that's why I am not used `isSetFlag`. correct me if I am wrong.

File: app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php (hunk 1 above)
Comment (srsathish92) [reply app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php]: Good point, I will update with suffix

File: app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php (hunk 3 above)
Comment (srsathish92) [reply app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php]: sure

File: app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php (hunk 2 above)
Comment (lbajsarowicz) [reply app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php]: There is no need to return anything from Test Method. Unit Tests are void.

In this case, what you are doing is state-based testing (based on the actions that are performed, not the result that is asserted).

Comment (srsathish92) [reply app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php]: @lbajsarowicz Thanks for feedback, I have update with assert. can you please review.

Comment (lbajsarowicz) [reply app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php]: Now is way better!

Comment (srsathish92) [reply app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php]: @lbajsarowicz Thanks for valuable feedback.

File: app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+    /**
+     * Data provider for isAvailableForCompare()
+     *
+     * @return array
+     */
+    public function isAvailableForCompareDataProvider(): array
+    {
+        return [
+            [1, true, ['is_in_stock' => true], false, true],
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php]: Try to remove as much magic numbers as you can. First column is `Status` - replace `1` with `\Magento\Catalog\Model\Product\Attribute\Source\Status::STATUS_ENABLED`, `2` with `\Magento\Catalog\Model\Product\Attribute\Source\Status::STATUS_DISABLED`. Of course - import that class.

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]

**Review by lbajsarowicz — APPROVED** [review]
Awesome. Thanks @srsathish92 for your contribution.

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-6477](https://jira.corp.magento.com/browse/ENGCOM-6477) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

File: app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php (hunk 4 above)
Comment (srsathish92) [reply app/code/Magento/Catalog/Test/Unit/ViewModel/Product/Checker/AddToCompareAvailabilityTest.php]: @lbajsarowicz Thanks for the feedback.

**Comment by engcom-Charlie** [comment]
:heavy_check_mark: QA Passed

**Pull request merged**